lib/secretsscanner/scan/scan_test.go

```diff
@@ -0,0 +1,192 @@
+/*
+ * Teleport
+ * Copyright (C) 2024  Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package scan
+
+import (
+	"context"
+	"os"
+	"path/filepath"
+	"sort"
+	"testing"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/uuid"
+	"github.com/stretchr/testify/require"
+	"golang.org/x/crypto/ssh"
+	cryptosshtestdata "golang.org/x/crypto/ssh/testdata"
+	"google.golang.org/protobuf/testing/protocmp"
+
+	accessgraphsecretsv1pb "github.com/gravitational/teleport/api/gen/proto/go/teleport/accessgraph/v1"
+	"github.com/gravitational/teleport/api/types/accessgraph"
+)
+
+var (
+	deviceID = uuid.NewString()
+)
+
+func TestNewScanner(t *testing.T) {
+	tests := []struct {
+		name    string
+		keysGen func(t *testing.T, path string) []*accessgraphsecretsv1pb.PrivateKey
+	}{
+		{
+			name:    "encrypted keys",
+			keysGen: writeEncryptedKeys,
+		},
+		{
+			name:    "unencrypted keys",
+			keysGen: writeUnEncryptedKeys,
+		},
+		{
+			name:    "encryptedKey without public key file",
+			keysGen: writeEncryptedKeyWithoutPubFile,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			dir := t.TempDir()
+
+			expect := tt.keysGen(t, dir)
+			s, err := NewScanner(ScannerConfig{Dirs: []string{dir}})
+			require.NoError(t, err)
+
+			keys := s.ScanPrivateKeys(context.Background(), deviceID)
+			var got []*accessgraphsecretsv1pb.PrivateKey
+			for _, key := range keys {
+				got = append(got, key.Key)
+			}
+
+			// Sort the keys by name for comparison.
+			sortPrivateKeys(expect)
+			sortPrivateKeys(got)
+
+			diff := cmp.Diff(expect, got, protocmp.Transform())
+			require.Empty(t, diff, "unexpected keys")
+
+		})
+	}
+}
+
+func sortPrivateKeys(keys []*accessgraphsecretsv1pb.PrivateKey) {
+	sort.Slice(keys, func(i, j int) bool {
+		return keys[i].Metadata.Name < keys[j].Metadata.Name
+	})
+}
+
+func writeEncryptedKeys(t *testing.T, dir string) []*accessgraphsecretsv1pb.PrivateKey {
+	t.Helper()
+	var expectedKeys []*accessgraphsecretsv1pb.PrivateKey
+	// Write encrypted keys to the directory.
+	for _, key := range cryptosshtestdata.PEMEncryptedKeys {
+		err := os.Mkdir(filepath.Join(dir, key.Name), os.ModePerm)
```

Comment:
nit: Write out the mode, the constant just makes it harder to understand what it is. It also seems needlessly permissive to me.

```suggestion
		err := os.Mkdir(filepath.Join(dir, key.Name), 0x644)
```

Same for others.